### PR TITLE
Opt out of /etc/nsswitch.conf for host and user lookup

### PR DESCRIPTION
**Pull request by neunhoef**
### Scope & Purpose

This is the next try to solve the problems described in
  https://arangodb.atlassian.net/wiki/spaces/DEV/pages/2188673090/etc+nsswitch.conf+incompatibility+with+glibc+static+executables

This PR uses the function `__nss_configure_lookup` to opt out of the
usage of /etc/nsswitch.conf for host name lookup and for user lookup.
It hard codes the lookup to /etc/hosts and normal DNS lookup, as well
as to /etc/passwd and /etc/group. There is a command line switch to
disable this opt out.

Note that this is an unofficial functionality of glibc. We only use
it when building under Linux and with glibc. Future versions might
not support it although this seems unlikely since glibc uses it in their
own tests.

This should prevent crashes of our static executables on systems which
provide a different glibc version than the one we build with.

- [*] :hankey: Bugfix

### Checklist

- [*] :book: CHANGELOG entry made
- [*] Backports
  - [*] Backport for 3.11: https://github.com/arangodb/arangodb/pull/21144
  - [*] Backport for 3.12.1: https://github.com/arangodb/arangodb/pull/21135

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [*] GitHub issue / Jira ticket: https://arangodb.atlassian.net/wiki/spaces/DEV/pages/2188673090/etc+nsswitch.conf+incompatibility+with+glibc+static+executables

### CODE REVIEW

**Review by goedderz — APPROVED** [review]

File: lib/ApplicationFeatures/ConfigFeature.cpp [hunk 1]
```diff
@@ -71,6 +77,13 @@ void ConfigFeature::collectOptions(std::shared_ptr<ProgramOptions> options) {
       new BooleanParameter(&_checkConfiguration),
       arangodb::options::makeDefaultFlags(arangodb::options::Flags::Uncommon,
                                           arangodb::options::Flags::Command));
+
+  options->addOption(
```
Comment (jsteemann) [inline lib/ApplicationFeatures/ConfigFeature.cpp:81]: As this option only has an effect when the executable was built with glibc, we could as well wrap the entire option into the ifdefs here as well.

Comment (neunhoef) [reply lib/ApplicationFeatures/ConfigFeature.cpp:81]: Yes, but then the observable behaviour of the executable (does the option exist) would change more between different builds (and OSes). I would prefer to leave it this way and have the option everywhere but render it meaningless outside of Linux/glibc builds.

File: lib/ApplicationFeatures/ConfigFeature.cpp [hunk 2]
```diff
@@ -71,6 +77,13 @@ void ConfigFeature::collectOptions(std::shared_ptr<ProgramOptions> options) {
...
+      "--honour-nsswitch",
```
Comment (Simran-B) [inline lib/ApplicationFeatures/ConfigFeature.cpp:82]: American English
```suggestion
      "--honor-nsswitch",
```

File: CHANGELOG [hunk 3]
```diff
@@ -1,6 +1,9 @@
 devel
 -----
 
+* Use __nss_configure_lookup to opt out of /etc/nsswitch.conf. Add option
+  --honour-nsswitch to cancel the opt out.
```
Comment (Simran-B) [inline CHANGELOG:5]:
```suggestion
  --honor-nsswitch to cancel the opt-out.
```

File: lib/ApplicationFeatures/ConfigFeature.cpp [hunk 4]
```diff
@@ -71,6 +77,13 @@ void ConfigFeature::collectOptions(std::shared_ptr<ProgramOptions> options) {
...
+      "Allow hostname lookup configuration via /etc/nsswitch.conf if on "
+      "Linux/glibc.",
+      new BooleanParameter(&_honourNsswitch),
```
Comment (Simran-B) [inline lib/ApplicationFeatures/ConfigFeature.cpp:85]:
```suggestion
      new BooleanParameter(&_honorNsswitch),
```

File: lib/ApplicationFeatures/ConfigFeature.cpp [hunk 5]
```diff
@@ -250,4 +263,34 @@ void ConfigFeature::loadConfigFile(std::shared_ptr<ProgramOptions> options,
...
+  // (like for example an older Ubuntu system or a Debian or RedHat system),
+  // then glibc tries to dynamically load a module which does not fit and the
+  // process crashes with a high likelyhood. To prevent this, we use the
+  // (undocumented) override function below. This has the consequence that the
+  // host name lookup will always just use /etc/hosts and normal DNS lookup. And
+  // username lookup will always just use /etc/passwd, regardless of the system
+  // configuration. There is an opt-out for this in form of the configuration
+  // option --honour-nsswitch. Use this only if you are running on a system
+  // without glibc installed, or with glibc version 2.39.0.
+  if (!_honourNsswitch) {
```
Comment (Simran-B) [inline lib/ApplicationFeatures/ConfigFeature.cpp:288]:
```suggestion
  // option --honor-nsswitch. Use this only if you are running on a system
  // without glibc installed, or with glibc version 2.39.0.
  if (!_honorNsswitch) {
```

File: lib/ApplicationFeatures/ConfigFeature.h [hunk 6]
```diff
@@ -53,7 +53,8 @@ class ConfigFeature final : public application_features::ApplicationFeature {
         }()},
         _file(configFilename),
         _progname(progname),
-        _checkConfiguration(false) {
+        _checkConfiguration(false),
+        _honourNsswitch(false) {
```
Comment (Simran-B) [inline lib/ApplicationFeatures/ConfigFeature.h:57]:
```suggestion
        _honorNsswitch(false) {
```

File: lib/ApplicationFeatures/ConfigFeature.h [hunk 7]
```diff
@@ -75,6 +78,8 @@ class ConfigFeature final : public application_features::ApplicationFeature {
   std::string _progname;
   std::vector<std::string> _defines;
   bool _checkConfiguration;
+  bool _honourNsswitch;  // If this is set to true, the internal override is
```
Comment (Simran-B) [inline lib/ApplicationFeatures/ConfigFeature.h:81]:
```suggestion
  bool _honorNsswitch;  // If this is set to true, the internal override is
```

**Review by Simran-B — CHANGES_REQUESTED** [review]

File: CHANGELOG (hunk 3 above)
Comment (Simran-B) [inline CHANGELOG:5]:
```suggestion
* Use __nss_configure_lookup to opt out of /etc/nsswitch.conf.
  Add the startup option --honor-nsswitch to cancel the opt-out.
```

**Review by jsteemann — APPROVED** [review]
LGTM

**Pull request merged**